### PR TITLE
Replace open() with smart_open() in notebooks. Fix #1789

### LINKED ISSUE #1789: open('aclImdb/alldata-id.txt', encoding='utf-8') doesn't work in Python 2.x
In file doc2vec-IMDB.ipynb, this statement is not supported in Python 2.x but is in Python 3.x

`with open('aclImdb/alldata-id.txt', encoding='utf-8') as alldata:
`

This works in Python 2.7:

`import codecs
`
`with codecs.open('aclImdb/alldata-id.txt', encoding='utf-8') as alldata:
`

### PR DESCRIPTION
**Pull request by sharanry**
Replace open() with smart_open() in notebooks to make it compatible with Python 3.x

Issue: #1789

### CODE REVIEW

**Comment by menshikh-iv** [comment]
Thanks for PR @sharanry, two questions:
- This replaces all `open` calls? 
- Are you test than notebooks works correctly (after replace)?

**Comment by sharanry** [comment]
i have replaced all open calls in notebooks directory.
what is the testing procedure? should i run the notebooks locally?

**Comment by menshikh-iv** [comment]
@sharanry yes

**Comment by sharanry** [comment]
@menshikh-iv Could you review this?

File: docs/notebooks/Corpora_and_Vector_Spaces.ipynb [hunk 1]
```diff
@@ -279,9 +279,10 @@
    },
    "outputs": [],
    "source": [
+    "from smart_open import smart_open\n",
     "class MyCorpus(object):\n",
     "    def __iter__(self):\n",
-    "        for line in open('datasets/mycorpus.txt'):\n",
+    "        for line in smart_open('datasets/mycorpus.txt'):\n",
```
Comment (menshikh-iv) [inline docs/notebooks/Corpora_and_Vector_Spaces.ipynb]: Better to explicitly add mode, i.e. `'rb'`, `'r'`, etc.

**Comment by sharanry** [comment]
@menshikh-iv I made the changes which you had asked. 
I have been trying to run the notebooks for the last couple of days. Some of the dependencies like nltk seem to be failing with the notebooks. I made sure they are all at the latest versions. Any suggestions?

**Comment by menshikh-iv** [comment]
@sharanry Also please merge `develop` from upstream to your branch, thanks.
That's all, let me know when you finished - I'll merge your PR.

**Comment by menshikh-iv** [comment]
Hey @sharanry, how about the run of notebooks?

**Comment by sharanry** [comment]
hey @menshikh-iv, sorry for the late reply, I havent been able to build the docker image, i tried `docker build -t gensim .` as given [here](https://github.com/RaRe-Technologies/gensim/blob/bd6db9a41baf219ecc4a1770cc21b01c8ff122e5/docker/README.md). I am getting this error.

gensim

**Comment by menshikh-iv** [comment]
@sharanry replace `locales=2.23-0ubuntu9` from https://github.com/RaRe-Technologies/gensim/blob/09672b369b2ec94a06581efd060be20662d7006c/docker/Dockerfile#L28 to  `locales=2.23-0ubuntu10`

**Comment by menshikh-iv** [comment]
@sharanry how is going? are you finished with running?

**Comment by sharanry** [comment]
@menshikh-iv Apologies for the delay. I am incurring multiple errors with the docker image. I am unable to set it up. 

Is there any alternative way I would run the notebooks?

**Comment by menshikh-iv** [comment]
@sharanry what's kind of errors, can you show it here?

alternative way - locally on your machine.

**Comment by sharanry** [comment]
gensim

I even tried with different network connections.

**Comment by sharanry** [comment]
Finally got it to work somehow.
Got a bunch of errors in Poincare Evaluations
poincareeval1
poincareeval2
poincareeval3
poincareeval4
There errors are causing even more errors

**Comment by sharanry** [comment]
Tensorboard visualizations doesn't have the required dataset
tboardvisual1

**Comment by sharanry** [comment]
Getting this error even after downloading `movie_plots.csv` from the link mentioned. Must be a problem with docker.

tboardvisual1

**Comment by menshikh-iv** [comment]
Big thanks @sharanry, need to fix notebooks first (unfortunately, this doesn't work correctly).

**Comment by sharanry** [comment]
@menshikh-iv I have yet to check all notebooks. I can help fixing them if possible.

**Comment by menshikh-iv** [comment]
@sharanry yes, it will be really nice, feel free to fix problems.

**Comment by sharanry** [comment]
@menshikh-iv Can we merge this?

**Comment by menshikh-iv** [comment]
@sharanry I want, but we need to run all of this notebooks first and check that all fine :(

**Comment by menshikh-iv** [comment]
@sharanry it was much harder than I thought
- Some notebooks need very long time for running
- I see many py3/py2 incompatibilities -> need to run notebook 2 or 3 times (27, 35, 36).
- Some problems with using (like - "let's go to download dataset from X" or using non-trivial dependencies).

I'll try to run, fix, merge this stuff when I have enough time for it (probably, not soon), sorry for waiting.

Also, please move commits from #1893 to current PR (anyway, I will not merge this until we fix it), also, this allows us to avoid painful merge-conflict with notebooks.

**Comment by menshikh-iv** [comment]
By discussion in #1964 (we'll rework most of the notebooks and provide small clear examples instead of most "long-running" stuff) + really hard to check (I spend a lot of time for it, but did not even check half, because of the need to test on 3 different python versions).

For this reason, I close current PR, sorry @sharanry, thanks for the work!

**Comment by piskvorky** [comment]
I see no reason to wait for some undefined future tests or refactorings just to change `open` to `smart_open`.

I propose merging this (useful) PR.

**Comment by menshikh-iv** [comment]
@piskvorky I close it because this needs a really long time to check that we didn't break something (and all works as expected) with this change.

**Comment by piskvorky** [comment]
`smart_open` is a drop-in replacement for `open`, so the checking is more checking of the rest of the notebook, not the changes here.

I agree it may not be a critical feature, but is a welcome one -- please merge.

File: docs/notebooks/lda_training_tips.ipynb [hunk 2]
```diff
@@ -67,7 +68,7 @@
     "    files = os.listdir(data_dir + yr_dir)\n",
     "    for filen in files:\n",
     "        # Note: ignoring characters that cause encoding errors.\n",
-    "        with open(data_dir + yr_dir + '/' + filen, errors='ignore') as fid:\n",
+    "        with smart_open(data_dir + yr_dir + '/' + filen, 'rb') as fid:\n",
```
Comment (piskvorky) [inline docs/notebooks/lda_training_tips.ipynb]: Bug (not in this PR, notebook already bad): use `os.path.join` for joining filesystem paths.

File: docs/notebooks/topic_coherence-movies.ipynb [hunk 3]
```diff
@@ -114,7 +115,7 @@
     "    # as well as pages about a single year.\n",
     "    # As a result, this preprocessing differs from the paper.\n",
     "    \n",
-    "    with open(os.path.join(data_dir, fname)) as f:\n",
+    "    with smart_open(os.path.join(data_dir, fname), 'rb') as f:\n",
```
Comment (piskvorky) [inline docs/notebooks/topic_coherence-movies.ipynb]: This looks like a bug. If you open the file in binary mode, operations like `split` and `lower` have a different meaning, compared to text.

I see the same problem in many places in this PR.

**Pull request merged**